[WIP] Replace $VIMRUNTIME with $NVIMRUNTIME

equalsraf opened the pull request:
- For compatibility plugins still need $VIMRUNTIME, but getting or
  settings VIMRUNTIME will use NVIMRUNTIME instead.

For #1924.

equalsraf commented on src/nvim/misc1.c:
```diff
@@ -2927,12 +2930,11 @@ char_u *vim_getenv(char_u *name, int *mustfree)
     return p;
   }
 
-  vimruntime = (STRCMP(name, "VIMRUNTIME") == 0);
```
This is using STRCMP() and I've kept it but _setenv uses STRICMP - shouldn't this be case insensitive too?

justinmk replied:
For NVIMRUNTIME, at least, I think we should do a case-sensitive comparison. Otherwise it's just more surface area to cover in tests, troubleshooting, and documentation.

Pyrohh replied:
Indeed, I've never seen environment variables interpreted case insensitively, and for good reason given how it doesn't adhere to the principle of least surprise, e.g. I'd be a little confused if `$ echo $VIMRUNTIME` showed nothing, but nvim was obeying `$ViMrUnTiMe`.

(I'm only mentioning this because I see some `STRICMP` calls still present in this PR)

equalsraf replied:
That is why I was asking, but I assumed I might be missing something so I left the comparisons exactly as I found them for fear of breaking something else.

Pyrohh replied:
:+1: I was just making sure.

On February 25, 2015 6:54:07 AM EST, Rui Abreu Ferreira <notifications@github.com> wrote:
>> @@ -2927,12 +2930,11 @@ char_u *vim_getenv(char_u *name, int
>*mustfree)
>>      return p;
>>    }
>>  
>> -  vimruntime = (STRCMP(name, "VIMRUNTIME") == 0);
>
>That is why I was asking, but I assumed I might be missing something so
>I left the comparisons exactly as I found them for fear of breaking
>something else.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/neovim/neovim/pull/1927/files#r25334308

justinmk commented on src/nvim/misc1.c:
```diff
@@ -2916,10 +2916,16 @@ char_u *vim_getenv(char_u *name, int *mustfree)
+    p = (char_u *)os_getenv("NVIM");
+  } else {
+    p = (char_u *)os_getenv((char *)name);
+  }
```
I haven't looked closely, but can you explain the intent of this new logic?

equalsraf replied:
In here, and in vim_getenv as well, whenever VIMRUNTIME is passed it is converted to NVIMRUNTIME i.e. VIMRUNTIME is never actually set and NVIMRUNTIME is set instead.

The same applies for VIM-NVIM.

fwalch commented on test/unit/os/env_spec.lua:
```diff
@@ -127,4 +129,60 @@ describe('env function', function()
+      eq(nil, os.getenv('VIMRUNTIME'))
+      eq(runtime, os.getenv('NVIMRUNTIME'))
+      -- Restore
+      vim_setenv('NVIMRUNTIME', old_runtime)
```
I think this will not be executed if a previous assertion fails. Maybe use a `setup`/`cleanup` function for the whole `describe` block to save/restore environment variables (if that's supported in busted)?

equalsraf commented:
@fwalch, confirmed there is are setup/teardown functions for busted, updated the tests.

fwalch commented on src/nvim/misc1.c:
```diff
@@ -3090,12 +3095,20 @@ static char_u *remove_tail(char_u *p, char_u *pend, char_u *name)
 void vim_setenv(char_u *name, char_u *val)
 {
-  os_setenv((char *)name, (char *)val, 1);
+  bool vim_runtime = (STRICMP(name, "VIMRUNTIME") == 0 || STRICMP(name, "NVIMRUNTIME") == 0);
```
`env_vimruntime`?

equalsraf replied:
yeah - either that, env_nvimruntime or env_runtime

fwalch commented on test/unit/os/env_spec.lua:
```diff
@@ -127,4 +129,66 @@ describe('env function', function()
+    local env_nvim
+
+    setup(function()
+      -- Save vars or '', nil is not acceptable and there is no unsetenv
```
Might be worth to implement an `os_unsetenv`. At least on Linux, there is an `unsetenv` function that could be used to implement that (`man 3 unsetenv`).

equalsraf replied:
I was thinking about the same thing, So far I was unable to find unsetenv for Windows.

fwalch replied:
I don't think it's supported on Windows, but we could wrap the workaround (set variable to "")
https://stackoverflow.com/questions/3205197/remove-environmental-variable-programmatically

justinmk replied:
related, though we don't need to worry about it for now: https://groups.google.com/d/msg/vim_dev/IIrDqVvdpOo/Q2nuImgZEtEJ

justinmk commented:
So what about `$MYVIMRC`? If we can VIMRUNTIME then we have to change that as well... and maybe others. Right?

kopischke commented:
@justinmk I’d think so too – `$VIMINIT` would be another candidate.

equalsraf commented:
Completely missed those two to be honest. The MYVIMRC might still be valid though, the user vimrc might work for both Vim and Neovim. For MYVIMRC we could use the same strategy, rename as MYNVIMRC on vim_setenv/getenv. 

I have not checked the VIMINIT case yet.

kopischke commented:
I think we need to look for a separate shell env var `$NVIMINIT`, as `$VIMINIT` is expressly designed to be set outside Vim to. One usage case (mine) is setting

```vim
let $MYVIMRC = "$XDG_CONFIG_HOME/vim/vimrc.vim" | execute "source" fnameescape($MYVIMRC)
```

to ensure Vim is looking for its config in the `$XDG_CONFIG/vim` directory; I could add a `has('nvim')` test of course, but I think being able to set a separate, Neovim specific variable, would be both more consistent and more straightforward.

equalsraf commented:
> as $VIMINIT is expressly designed to be set outside Vim to

Indeed, it is only used in main.c with `process_env()` and  it shouldn't be used in vimscript. It seems we can just rename VIMINIT to NVIMINIT there.

equalsraf committed:
Rename VIMINIT as NVIMINIT

- Use NVIMINIT environment var instead of VIMINIT
- Update documentation

kopischke commented:
Should we fall back to `$VIMINIT` when `$NVIMINIT` is not set? That’‚ what we do for `.nvimrc`→ `.vimrc`, correct?

justinmk commented:
If we fall back then it defeats the purpose...

kopischke commented:
> If we fall back then it defeats the purpose...

That’s why I was asking about how we handle a missing `.nvimrc` – I was thinking of keeping this consistent, i.e. if we don’t fall back to `.vimrc`, not falling back to `$VIMINIT` makes sense, and *vice versa*.

jszakmeister commented:
@kopischke We don't fallback.  I believe we used to at one point, IIRC, but I'm pretty sure that's gone for a while now.

kopischke commented:
> We don't fallback.

Then I agree **not** falling back is the right course fore `$NVIMINIT` too. As an aside, @justinmk, is there a technical reason the relevant help docs aren’t yet updated for these differences (apart from the `vim_differences.txt` PR)? Merge compatibility with upstream patches or something like that? Or would it make sense to submit a PR?

justinmk commented:
> is there a technical reason the relevant help docs aren’t yet updated

No reason. We've updated a lot of documentation, but it turns out there is a lot more :)

kopischke commented:
> No reason. We've updated a lot of documentation, but it turns out there is a lot more :)

I’ll submit a documentation PR once this gets merged, then :).

equalsraf commented:
Rebased, updated tests for the new vim_getenv bool parameter. The oldstests are failing on mac+gcc though

tarruda commented:
@equalsraf test 61 has been failing randomly for some time

tarruda commented:
If you restart the build I think it will pass

equalsraf commented:
Yup, it passed.

Pyrohh commented on src/nvim/eval.c:
```diff
@@ -1754,10 +1754,10 @@ ex_let_one (
           if (STRICMP(name, "HOME") == 0)
             init_homedir();
-          else if (didset_vim && STRICMP(name, "VIM") == 0)
+          else if (didset_vim && (STRICMP(name, "VIM") == 0 || STRICMP(name, "NVIM") == 0))
```
Why is `STRICMP` being used over `STRCMP`? I don't mean to come across as blunt, I'm just curious because of what was discussed here: https://github.com/neovim/neovim/pull/1927#discussion_r23903426

equalsraf replied:
Honestly I was afraid to change it, since I have no idea why it is set to STRICMP.

ZyX-I commented:
I now want to know, how can I set $VIMRUNTIME or $VIM? I do not like the idea of code like

    let $VIMRUNTIME='/tmp/test'
    !vim -u NORC --cmd '…'

doing *not* what I want. And am highly against such untransparent changes.

Vim already has many black corners. What you are doing is adding yet another one. I would never accept such a change:

    let ${VARNAME} = 'foo'

*does* set the environment variable named `{VARNAME}`. *Not* sets the environment variable named `VARNAME`, unless ….

    echo ${VARNAME}

does echo value of the environment variable named `{VARNAME}`, *not* echoes it *unless*…

Never saw a language which has such unclear behaviour in its core. I am hoping this PR will not ever be merged until relevant changes in `vim_[sg]etenv` are removed.

equalsraf commented:
```
let $VIMRUNTIME='/tmp/test'
!vim -u NORC --cmd '…'
```
> Vim already has many black corners. What you are doing is adding yet another one. I would never accept such a change:

@ZyX-I  valid points, that  example wont work as intended, but right now we can't do this either

```
!vim --cmd '…'
```
since vim will get the incompatible Neovim runtime from the default $VIMRUNTIME=/usr/share/nvim. 

We also can't stop using $VIMRUNTIME and $VIM in VimL because lots of scripts depend on those.

Even If we assume that scripts are well behaved i.e. they only `let  $VIMRUNTIME=` when calling vim and restore it afterwards, maybe we could

1. remove the obscure VIM->NVIM mapping from `vim_setenv()` i.e. `let ` works as expected
2. Change vim_getenv() to map $VIM -> $NVIM **only if**  $VIM were never set from `vim_setenv()` (likewise for NVIMRUNTIME) i.e. only gets those vars if they were set from VimL

But this is also problematic it the script actually intended to affect VIMRUNTIME for neovim, so it is only marginally less obscure than the implementation in this PR.

@ZyX-I I'm not seeing other ways to simultaneously avoid conflicts with vim and retain compatibility with existing VimL scripts.

ZyX-I commented:
> @ZyX-I I'm not seeing other ways to simultaneously avoid conflicts with vim and retain compatibility with existing VimL scripts.

I think that suggested price is too much for this goal. Even if there is no other way I do not want to pay it. Note that these variables are almost never used in plugins, mostly in vimrc (e.g. to source plugins).

I am also thinking that you are solving a non-problem because vim from nvim is normally not used, or only used by such strange people as developers who know how to cope with problems. Thus just running `let $VIMRUNTIME = $NVIMRUNTIME` at startup is sufficient most of time, adding ShellVarChanged autocmd (note: core will receive only *event*; autocmd doing changes itself should be in core plugins) which will do this thing when setting NVIMRUNTIME.

In any case you don’t need to do anything with setting the environment variables at all because this is far, far less common: out of [71367](https://github.com/search?l=vim&p=3&q=VIMRUNTIME&type=Code&utf8=%E2%9C%93) `VIMRUNTIME` matches in VimL files [40](https://github.com/search?l=vim&q=%22let+%24VIMRUNTIME%22&ref=searchresults&type=Code&utf8=%E2%9C%93) are `let $VIMRUNTIME` exact matches and [6740](https://github.com/search?l=vim&q=%22so+%24VIMRUNTIME%22&ref=searchresults&type=Code&utf8=%E2%9C%93) + [15 419](https://github.com/search?l=vim&q=%22source+%24VIMRUNTIME%22&ref=searchresults&type=Code&utf8=%E2%9C%93) are `so $VIMRUNTIME` + `source $VIMRUNTIME` exact matches.

equalsraf commented:
> I think that suggested price is too much for this goal. Even if there is no other way I do not want to pay it. Note that these variables are almost never used in plugins, mostly in vimrc (e.g. to source plugins).

vimrc and in runtime/ VimL most likely yes.

> I am also thinking that you are solving a non-problem because vim from nvim is normally not used, 

The two original issues that led here were:

1. nvim called git (from some plugin), git launched $EDITOR=vim, vim used the value from $VIMRUNTIME which was set by nvim to its runtime
2. (from #1924) user had a custom $VIMRUNTIME in bashrc (for vim) and neovim tried to use it, naturally it failed

which is why I was trying to allow $VIMRUNTIME to remain unchanged (by neovim)

ZyX-I commented:
These two cases are solved *just* by making nvim respect only `NVIMRUNTIME` and set only it. I do not see why you think you need to taint `vim_setenv`/`vim_getenv` to solve these issues.

justinmk commented:
I'm leaning towards @ZyX-I concerns here. Avoiding surprising behavior is more important than backwards compatibility--at least in this case, but perhaps generally as well.

ZyX-I commented:
These cases are solved *just* by making nvim respect only `NVIMRUNTIME`. I see that you need to taint `vim_[sg]etenv` to make nvim compatible with scripts while this change is in effect and I do not see how you can solve this clearly, but second case is 90% fine (matchit is not neovim-specific) and absolutely fine if you `let $VIMRUNTIME=$NVIMRUNTIME` at startup and first case can be solved by altering the vimrc to set `VIMRUNTIME` value to default if it detects it be equal with `NVIMRUNTIME` and is absolutely fine if you do *not* do `let $VIMRUNTIME=$NVIMRUNTIME` at startup.

equalsraf committed:
Set $VIM=$NVIM and $VIMRUNTIME=$NVIMRUNTIME

- Avoid accidentaly loading the Vim runtime, use NVIM* environment variables
  instead
- Updated tests to use $NVIMRUNTIME

equalsraf commented:
Ok, for starters I've reverted the vim_[g/s]etenv changes. Changed `early_init()`to set $VIM=$NVIM and $VIMRUNTIME=$NVIMRUNTIME at startup.

> and first case can be solved by altering the vimrc to set VIMRUNTIME value to default if it detects it be equal with NVIMRUNTIME and is absolutely fine if you do not do let $VIMRUNTIME=$NVIMRUNTIME at startup.

Neovim does keep track of `let  $VIM` in the `didset_vim` global var (and didset_vimruntime). Would it be  possible (maybe, have not checked jobstart bits) to restore the original values for VIM/RUNTIME during calls to `os_system`?

Pyrohh commented on runtime/doc/options.txt:
```diff
@@ -128,7 +128,7 @@ A few special texts:
 	Last set from environment variable ~
-		Option was set from an environment variable, $VIMINIT,
+		Option was set from an environment variable, $NVIMINIT,
 		$GVIMINIT or $EXINIT.
```
GVIMINIT isn't actually checked, so could you remove it? Excuse me if it's out of the scope of this PR.

equalsraf committed:
Remove reference to GVIMINIT from docs

splinterofchaos commented:
@equalsraf could you squash the "Add neq import to env_spec.lua" with the "Implement os_unsetenv()" commit? I'd like to cherry pick it for #2331.

equalsraf commented:
@splinterofchaos squashed the neq import into the first commit.

splinterofchaos commented:
Thanks!

bfredl commented on runtime/doc/starting.txt:
```diff
@@ -545,13 +545,13 @@ accordingly.  Vim proceeds in this order:
 	For the Macintosh the $VIMRUNTIME/macmap.vim is read.
 
-	  *VIMINIT* *.vimrc* *_vimrc* *EXINIT* *.exrc* *_exrc* *$MYVIMRC*
+	  *NVIMINIT* *.vimrc* *_vimrc* *EXINIT* *.exrc* *_exrc* *$MYVIMRC*
```
while changing this line, perhaps fix .vimrc to .nvimrc here as well?

Pyrohh replied:
This breaks many tags. If you go to `runtime/doc`, run `make html`, a bunch of warnings should pop up (assuming you didn't do `-j2`, which causes no warnings to pop up for me).

Pyrohh replied:
Well there's plenty more `.vimrc` references than just those pertaining to this tag, and it's not directly related to this PR so I don't think it should be done here.

bfredl replied:
Ok, then probably `.nvimrc` tag should be added instead, and `nvimrc` tag after `vimrc` above  as `:help nvimrc` fails right now, but yes this unrelated to this PR.

bfredl commented on src/nvim/main.c:
```diff
@@ -1789,7 +1804,7 @@ static void source_startup_scripts(mparm_T *parmp)
-     * - environment variable VIMINIT
+     * - environment variable NVIMINIT
      * - user vimrc file (~/.vimrc)
      * - second user vimrc file ($VIM/.vimrc for Dos)
```
`$NVIM/.nvimrc` ?

equalsraf committed:
Add unit tests for $NVIM/$NVIMRUNTIME

equalsraf commented:
Travis succeeded but quickbuild is failling, seems to be setting $VIMRUNTIME somewhere.

Pyrohh commented:
I don't think it's much to worry about for now, as everything else seems to be failing :D

justinmk commented on src/nvim/main.c:
```diff
@@ -1808,7 +1823,7 @@ static void source_startup_scripts(mparm_T *parmp)
      * The first that exists is used, the rest is ignored.
      */
-    if (process_env((char_u *)"VIMINIT", TRUE) != OK) {
+    if (process_env((char_u *)"NVIMINIT", TRUE) != OK) {
```
If we don't support VIMINIT then we definitely should also remove support for EXINIT. VIMINIT exists only as an alternative to EXINIT for users who want separate Vi and Vim `INIT`s :)

justinmk commented on src/nvim/main.c:
```diff
@@ -1907,7 +1922,7 @@ static void main_start_gui(void)
 process_env (
     char_u *env,
-    int is_viminit             /* when TRUE, called for VIMINIT */
+    int is_nviminit             /* when TRUE, called for NVIMINIT */
```
Internal identifiers should continue to use `vim` as a generic "protocol" name. `is_viminit` is fine.

justinmk commented on test/unit/env_nvim2_spec.lua:
```diff
@@ -0,0 +1,23 @@
+  env.os_setenv('VIM', 'VIM_VALUE', 1)
+  env.os_setenv('VIMRUNTIME', 'VIMRUNTIME_VALUE', 1)
+
+  helpers.vim_init()
```
Is this why the tests need to be separate? Or can they be combined to 1 env_nvim_spec.lua file? I am also wondering why there are nested `describe()` blocks with no `it()` specs. Suggestion:

```lua
local helpers = require("test.unit.helpers")
local cimport = helpers.cimport
local env = cimport('./src/nvim/os/os.h')
local eq = helpers.eq
local neq = helpers.neq

describe('$NVIM/$NVIMRUNTIME', function()
  before_each(function()
    env.os_unsetenv('NVIM')
    env.os_unsetenv('NVIMRUNTIME')
    env.os_unsetenv('VIM')
    env.os_unsetenv('VIMRUNTIME')

    helpers.vim_init()
  end)

  it('$VIM is cleared if $NVIM is unset', function()
    env.os_setenv('VIM', 'VIM_VALUE', 1)
    neq(os.getenv('VIM'), 'VIM_VALUE')
  end)
  it('$VIMRUNTIME is cleared if $NVIMRUNTIME is unset', function()
    env.os_setenv('VIMRUNTIME', 'VIMRUNTIME_VALUE', 1)
    neq(os.getenv('VIMRUNTIME'), 'VIMRUNTIME_VALUE')
  end)

  it('$VIM is set to $NVIM', function()
    env.os_setenv('NVIM', 'NVIM_VALUE', 1)
    eq(os.getenv('VIM'), os.getenv('NVIM'))
  end)
  it('$VIMRUNTIME is set to $NVIMRUNTIME', function()
    env.os_setenv('NVIMRUNTIME', 'NVIMRUNTIME_VALUE', 1)
    eq(os.getenv('VIMRUNTIME'), os.getenv('NVIMRUNTIME'))
  end)
end)


```

justinmk commented on src/nvim/main.c:
```diff
@@ -145,6 +145,21 @@ void early_init(void)
+  }
+  const char *env_nvim = os_getenv("NVIM");
+  if (env_nvim) {
+    os_setenv("VIM", env_nvim, true);
```
Let's leave this out for now, and instead just change the references (there's only ~6) in option.c, eval.c, ex_cmds.c, and env.c.

If the backwards compatibility thing turns out to be a nightmare, *then* we can consider setting `$VIM`. But it should be a last resort.

equalsraf replied:
Wouldn't that break scripts all over? i.e. if VIM is not set and we don't set it in `vim_getenv()` $VIM will never hold the value expected by scripts.

justinmk replied:
Yes, I just did a GH search for `$VIMRUNTIME` and it yielded >1 million results.

justinmk commented on src/nvim/main.c:
```diff
@@ -145,6 +145,21 @@ void early_init(void)
+  // or unset them to avoid accidentaly using the Vim runtime
+  const char *env_nvimruntime = os_getenv("NVIMRUNTIME");
+  if (env_nvimruntime) {
+    os_setenv("VIMRUNTIME", env_nvimruntime, true);
```
Instead of this let's just replace the references to VIMRUNTIME in the nvim source (for reasons mentioned in the other comment). It's a bit more work, but it's the right thing to do.

justinmk commented:
After reviewing https://github.com/neovim/neovim/issues/1924 and thinking about the costs and benefits, I again think it would be best to simply do a sanity check on VIMRUNTIME and show a warning. That solves #1924 and avoids these other costs. 

If we solve #1924 by changing environment variables, then it's a loss for predictability and usability. 

- Users will be confused and will wonder what *other* variables nvim is changing without permission. Users lose trust in the application.
- Anything that inherits the nvim environment potentially needs to fix VIMRUNTIME for a vi script or whatever.

We can trivially detect whether VIMRUNTIME is valid by looking for script(s) that only exist in Neovim.

equalsraf commented:
> We can trivially detect whether VIMRUNTIME is valid by looking for script(s) that only exist in Neovim.

Something like this then

```diff
diff --git a/src/nvim/main.c b/src/nvim/main.c
index af6e49c..e556f81 100644
--- a/src/nvim/main.c
+++ b/src/nvim/main.c
@@ -145,6 +145,19 @@ void early_init(void)
 {
   handle_init();
 
+  // Check if $VIMRUNTIME points to a valid
+  // Neovim runtime path, look for doc/nvim_intro.txt
+  const char *runtime = os_getenv("VIMRUNTIME");
+  if (runtime && runtime[0]) {
+    char rt_testpath[MAXPATHL] = "";
+    STRNCPY(rt_testpath, runtime, MAXPATHL);
+    append_path(rt_testpath, "doc", MAXPATHL);
+    append_path(rt_testpath, "nvim_intro.txt", MAXPATHL);
+    if (!os_file_exists((char_u *)rt_testpath)) {
+      EMSG2(_("EXX: Invalid Neovim runtime from $VIMRUNTIME=%s"), runtime);
+    }
+  }
+
```

justinmk commented:
@equalsraf Yes, exactly that. Though it probably can be done well after `early_init()`.

fwalch commented:
Closing this, as we don't want to replace VIMRUNTIME anymore. See #2977.
